Skeleton Codec Support Page (#188)
jellyfin/jellyfin-docs: general/clients/codecs.md

 
## [Video Compatibility](https://en.wikipedia.org/wiki/Comparison_of_video_container_formats "Wikipedia's video codec tables")

||WebOS|Android|AndroidTV|Kodi|[Roku](https://developer.roku.com/docs/specs/streaming.md)|
|:---:|:---:|:---:|:---:|:---:|:---:|
|[H.262/Xvid](https://en.wikipedia.org/wiki/MPEG-4_Part_2)|✅|✅|✅|✅|✅|
|[H.262/DivX](https://en.wikipedia.org/wiki/DivX)|✅|✅|✅|✅|✅|
|[H.264/AVC](https://caniuse.com/#feat=mpeg4 "H264 Browser Support Reference")|🔶<sup>3</sup>|✅|✅|✅|✅|
|[H.265/HEVC](https://caniuse.com/#feat=hevc "HEVC Browser Support Reference")|❌<sup>1</sup>|🔶<sup>2</sup>|❌|✅|❌|

<sup>1</sup>HEVC support is potentially possible by offloading to the OS. *untested*

> <sup>2</sup>Android playback is currently broken. Client reports that HEVC is supported and attempts to Directstream it.

Reviewer: ```suggestion
<sup>2</sup>Android playback is currently broken. Client reports that HEVC is supported and attempts to Direct Stream.
```

Author: hm... depending on the file and os, it could be direct play. semantics, but food for thought

||WebOS|Android|AndroidTV|Kodi|Roku
|:---:|:---:|:---:|:---:|:---:|:---:
|MP3|🔶<sup>1</sup>|✅|✅|✅|✅
|AAC|✅|✅|✅|✅|✅
|AC3|✅|✅|✅|✅|✅
|EAC3<sup>2</sup>|✅|✅|✅|✅|✅
|VORBIS|❌|✅|✅|✅|✅

<sup>1</sup>MP3 Mono is incorrectly reported as unsupported.

<sup>2</sup>Only EAC3 2.0 has been tested.

## [Subtitle Compatibility](https://en.wikipedia.org/wiki/Comparison_of_video_container_formats#Subtitle/caption_formats_support "Wikipedia's subtitle codec tables")

Subtiles can be a subtle issue for transcoding. Containers have a limited number of subtitles that are supported. If subtitles need to be transcoded, it will happen one of two ways. They can be converted into another supported format (text-based subtitles) or burned into the video (image/lossless based and ASS based) due to the subtitles transcoding not being supported. This is the most intenstive method of transcoding due to two transcodings happening at once; applying the subtitle layer on top of the video layer. 

||Format|TS|MP4|MKV|AVI|
|:---:|:---:|:---:|:---:|:---:|:---:|

> |SubRip Text (SRT)|Formatted Text|❌|🔶|✅|🔶|

Reviewer: ```suggestion
|SubRip Text (SRT)|Format|❌|🔶|✅|🔶|
```

Author: Format or Text? I felt like it was text, but I was follow wikipedia's table. I haven't check any compatibility yet.

|ASS/SSA|Formatted Text|❌|❌|✅|🔶|
|VobSub|Picture|❌|✅|✅|🔶|
|DVB-SUB|Picture|✅|❌|✅|❌|
|MP4TT/TXTT|XML|❌|✅|❌|❌|

> Note: ASS Subtitles are only supported by mkv files. Mkv files can't natively be streamed therefore ASS subtitles will always inherently be burned into the video. This is not a limitation of JF. 

Reviewer: maybe use the `<sus>` thing here like you did above, or at least the formatted note element

Author: i just added the subtitles table, i haven't had a chance yet. there's so many rules with subtitles, a third of my notes are on subs alone lol

## [Container Compatibility](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers)

> If the container is unsupported, this will result in remuxing. The video and audio codec will remain intact, but wrapped in a container that is supported. This is the least intensive process. Most video containers will be remuxed to use the hls streaming protocol and ts containers. Remuxing shouldn't be a concern even for a Rpi3.

Reviewer: ```suggestion
If the container is unsupported, this will result in remuxing. The video and audio codec will remain intact, but wrapped in a container that is supported. This is the least intensive process. Most video containers will be remuxed to use the hls streaming protocol and ts containers. Remuxing shouldn't be a concern even for a RPi3.
```

||WebOS|Android|AndroidTV|Kodi|Roku
|:---:|:---:|:---:|:---:|:---:|:---:
|[mp4](https://en.wikipedia.org/wiki/MPEG-4_Part_14)<sup>1</sup>|✅|✅|✅|✅|✅
|[MKV](https://en.wikipedia.org/wiki/Matroska)<sup>2, 3</sup>|❌|✅|🔶|✅|🔶

> |[ts](https://en.wikipedia.org/wiki/MPEG_transport_stream)|✅|✅|✅|✅|✅

Reviewer: There's a bit of inconsistent capitalization with the formats in the document.

Author: yeah, it's still a little rough. I need to go through and pick a standard. Maybe follow the official capitalization for each like webM

---
uid: codecs
title: Codec Compatibility
---

 # [Codec Tables](https://en.wikipedia.org/wiki/List_of_codecs "Wikipedia's list of all codecs")
 

>  The goal is to DirectPlay all media. This means the container, video, audio and subtitles are all compatible with the client. DirectStream will occur if the audio, container or subtitles happen to not be supported. Subtitles can be tricky because they can cause DirectStreaming (subtitles are remuxed) or video transcoding (burning in subtitles) to occur. If the video codec is unsupported, this will result in video transcoding. This is the most intensive CPU component of transcoding. Decoding is less intensive than encoding.

Reviewer: ```suggestion
 The goal is to Direct Play all media. This means the container, video, audio and subtitles are all compatible with the client. Direct Stream will occur if the audio, container or subtitles happen to not be supported. Subtitles can be tricky because they can cause Direct Stream (subtitles are remuxed) or video transcoding (burning in subtitles) to occur. If the video codec is unsupported, this will result in video transcoding. This is the most intensive CPU component of transcoding. Decoding is less intensive than encoding.
```

 
## [Video Compatibility](https://en.wikipedia.org/wiki/Comparison_of_video_container_formats "Wikipedia's video codec tables")

||Chrome|Firefox|Safari|Android|AndroidTV|Kodi|[Roku](https://developer.roku.com/docs/specs/streaming.md)|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|[MPEG-4 Part 2/SP](https://en.wikipedia.org/wiki/DivX)|❌|❌|❌|❌|❌|✅|✅|
|[MPEG-4 Part 2/ASP](https://en.wikipedia.org/wiki/MPEG-4_Part_2#Advanced_Simple_Profile_(ASP))|❌|❌|❌|❌|❌|✅|✅|
|[H.264 8Bit](https://caniuse.com/#feat=mpeg4 "H264 Browser Support Reference")|✅|✅|✅|✅|✅|✅|✅|
|[H.264 10Bit](https://caniuse.com/#feat=mpeg4 "H264 Browser Support Reference")|✅|❌|❌|✅|✅|✅|✅|
|[H.265](https://caniuse.com/#feat=hevc "HEVC Browser Support Reference")|❌|❌|❌<sup>1</sup>|🔶<sup>2</sup>|❌|✅|❌|

> <sup>1</sup>HEVC support is potentially possible by offloading to the OS. *untested*

Reviewer: ```suggestion
<sup>1</sup>HEVC support is potentially available by offloading to the operating system, but this has not been tested.
```

|:---:|:---:|:---:|:---:|:---:|:---:|
|H.262|MPEG-2 Video|MPEG4 SP/ASP|H.264|MP4 Container<sup>2</sup>|H.265|
|DVD-Video||DivX|MPEG-4 AVC||HEVC|
|||DX50||||

<sup>1</sup>[MPEG-4 Part-2 vs Part-10](https://www.afterdawn.com/glossary/term.cfm/mpeg_4_part_10)

<sup>2</sup>[MPEG-4 Part 17: MP4TT Subtitles](https://en.wikipedia.org/wiki/MPEG-4_Part_17)

## [Audio Compatibility](https://en.wikipedia.org/wiki/Comparison_of_video_container_formats#Audio_coding_formats_support "Wikipedia's audio codec tables")

If the audio codec is unsupported or incompatible (such as playing a 5.1 channel stream on a stereo device), the audio codec must be transcoded. This is not nearly as intensive as video coding.

||Chrome|Firefox|Safari|Android|AndroidTV|Kodi|Roku|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
FLAC|✅|❌|✅|✅|✅|✅|✅|
|MP3|🔶<sup>1</sup>|🔶|✅|✅|✅|✅|✅|
|AAC|🔶<sup>2</sup>|🔶|✅|✅|✅|✅|✅|
|AC3|✅|❌|✅|✅|✅|✅|✅|
|EAC3<sup>3</sup>|✅|✅|✅|✅|✅|✅|✅|
|VORBIS|❌|✅|✅|✅|✅|✅|✅|
|DTS<sup>4</sup>|❌|❌|❌|✅|✅|✅|✅|

<sup>1</sup>MP3 Mono is incorrectly reported as unsupported and will transcode to AAC.

> <sup>2</sup> AAC is incorrectly reported as unsupported and will transcode to MP3.

Reviewer: ```suggestion
<sup>2</sup>AAC is incorrectly reported as unsupported and will transcode to MP3.
```

<sup>3</sup>Only EAC3 2.0 has been tested.

> <sup>4</sup> Only DTS Mono has been tested.

Reviewer: ```suggestion
<sup>4</sup>Only DTS Mono has been tested.
```

## [Subtitle Compatibility](https://en.wikipedia.org/wiki/Comparison_of_video_container_formats#Subtitle/caption_formats_support "Wikipedia's subtitle codec tables")

Subtiles can be a subtle issue for transcoding. Containers have a limited number of subtitles that are supported. If subtitles need to be transcoded, it will happen one of two ways. They can be converted into another supported format (text-based subtitles) or burned into the video (image/lossless based and ASS based) due to the subtitles transcoding not being supported. This is the most intenstive method of transcoding due to two transcodings happening at once; applying the subtitle layer on top of the video layer.

||Format|TS|MP4|MKV|AVI|
|:---:|:---:|:---:|:---:|:---:|:---:|

> |SubRip Text (.srt)|Text|❌|🔶|✅|🔶|

Reviewer: ```suggestion
|SubRip Text (SRT)|Text|❌|🔶|✅|🔶|
```

|ASS/SSA<sup>1</sup>|Formatted Text|❌|❌|✅|🔶|
|VobSub|Picture|❌|✅|✅|🔶|

> |DVB-SUB [(.sub/.idx)](https://forum.videohelp.com/threads/261451-Difference-between-SUB-and-IDX-file)|Picture|✅|❌|✅|❌|

Reviewer: ```suggestion
|DVB-SUB [(SUB or IDX)](https://forum.videohelp.com/threads/261451-Difference-between-SUB-and-IDX-file)|Picture|✅|❌|✅|❌|
```

Author: No, idx is the index file for sub. See if you can find link for it and add it, if not I'll grab it real quick. My notes are in my upstream branch. If you can't find it, don't worry about it, I got it ;)

Author: i dont think it can be an or statement. It looks like they're both needed. I'm not sure just because I've never looked at external sub files that werent embedded in the media file

Author: I'll change it to a plus sign? not sure how i want to format that. why can't it just be easy? lol
```suggestion
|DVB-SUB [(SUB + IDX)](https://forum.videohelp.com/threads/261451-Difference-between-SUB-and-IDX-file)|Picture|✅|❌|✅|❌|
```

Author: Change it to a plus sign? i just love formatting >.>

|MP4TT/TXTT|XML|❌|✅|❌|❌|
|PGSSUB|Picture|❌|❌|✅|❌|

> 

Reviewer: extra newline

> <sup>1</sup>ASS Subtitles are only supported by mkv files. Mkv files aren't supported by Firefox.They will always inherently be burned into the video. This is not a limitation of JF. 

Reviewer: ```suggestion
<sup>1</sup>ASS Subtitles are only supported by mkv files. MKV files aren't supported by Firefox.They will always inherently be burned into the video. This is not a limitation of Jellyfin.
```

Reviewer: watch out for those spaces at the end of sentences ;) just noticed this one haha

> <sup>1</sup>ASS Subtitles are only supported by MKV files. MKV files aren't supported by Firefox.They will always inherently be burned into the video. This is not a limitation of Jellyfin.

Reviewer: ```suggestion
<sup>1</sup>ASS Subtitles are only supported by MKV files. MKV files aren't supported by Firefox. They will always inherently be burned into the video. This is not a limitation of Jellyfin.
```

## [Container Compatibility](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers)

> If the container is unsupported, this will result in remuxing. The video and audio codec will remain intact, but wrapped in a container that is supported. This is the least intensive process. Most video containers will be remuxed to use the hls streaming protocol and ts containers. Remuxing shouldn't be a concern even for a RPi3.

Reviewer: ```suggestion
If the container is unsupported, this will result in remuxing. The video and audio codec will remain intact, but wrapped in a supported container. This is the least intensive process. Most video containers will be remuxed to use the HLS streaming protocol and TS containers. Remuxing shouldn't be a concern even for an RPi3.
```

||Browser|Android|AndroidTV|Kodi|Roku
|:---:|:---:|:---:|:---:|:---:|:---:
|[MP4](https://en.wikipedia.org/wiki/MPEG-4_Part_14)<sup>1</sup>|✅|✅|✅|✅|✅
|[MKV](https://en.wikipedia.org/wiki/Matroska)<sup>2, 3</sup>|❌|✅|🔶|✅|🔶
|[TS](https://en.wikipedia.org/wiki/MPEG_transport_stream)|✅|✅|✅|✅|✅

> |[webM](https://en.wikipedia.org/wiki/WebM)<sup>3</sup>|||||

Reviewer: ```suggestion
|[WEBM](https://en.wikipedia.org/wiki/WebM)<sup>3</sup>|||||
```

> |[webM](https://en.wikipedia.org/wiki/WebM)<sup>3</sup>|||||

Reviewer: ```suggestion
|[WebM](https://en.wikipedia.org/wiki/WebM)<sup>3</sup>|||||
```

<sup>1</sup>MP4 containers are one of the few containers that will not remux.

<sup>2</sup>MKV containers can hold nearly any codec, but are not compatible with streaming in Firefox and will remux.

> <sup>3</sup>MKV containers are improperly labeled as webM in Firefox during playback.

Reviewer: ```suggestion
<sup>3</sup>MKV containers are improperly labeled as WEBM in Firefox during playback.
```

Author: hm.... it should actually be WebM

> <sup>3</sup>MKV containers are improperly labeled as webM in Firefox during playback.

Reviewer: ```suggestion
<sup>3</sup>MKV containers are improperly labeled as WebM in Firefox during playback.
```
